### PR TITLE
CI: -Werror for host compilers

### PR DESCRIPTION
**Pull request by ax3l**
## Summary

Add more warning coverage in CI for CMake builds.

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [ ] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] are described in the proposed changes to the AMReX documentation, if appropriate

### CODE REVIEW

**Comment by ax3l** [comment]
@mic84 I think about adding something like this block, guarded by being the root project in a super build, to our CMake logic to be on par with the warning coverage in downstream projects (i.e. WarpX) and GNUmake. What do you think?

https://github.com/ECP-WarpX/WarpX/blob/51f5af9f1950209948c0e91ff9e2882d9b9c5fed/cmake/WarpXFunctions.cmake#L82-L113

File: Src/Base/AMReX_GpuAsyncArray.H [hunk 1]
```diff
@@ -22,6 +22,19 @@ extern "C" {
 #endif
 
 namespace amrex {
+
+namespace type_traits {
+#if __cplusplus < 202002L
```
Comment (ax3l) [inline Src/Base/AMReX_GpuAsyncArray.H]: Hah, GCC 10.1 identifies as C++17 if compiled with `-std=c++2a` but already adds the C++20 deprecation notice ;)

**Comment by mic84** [comment]
> @mic84 I think about adding something like this block, guarded by being the root project in a super build, to our CMake logic to be on par with the warning coverage in downstream projects (i.e. WarpX) and GNUmake. What do you think?
> 
> https://github.com/ECP-WarpX/WarpX/blob/51f5af9f1950209948c0e91ff9e2882d9b9c5fed/cmake/WarpXFunctions.cmake#L82-L113

@ax3l I am in favour of adding extra warnings but I would prefer not to do it via ``CMAKE_CXX_FLAGS``

**Comment by ax3l** [comment]
@mic84 sounds good, let's to this in a follow-up PR.

**Review by mic84 — APPROVED** [review]

**Review by WeiqunZhang — APPROVED** [review]

**Pull request merged**